### ISSUE TITLE
Allow user to update metrics in Tensorboard SummaryWriter.add_hparam

### ISSUE THREAD

**Issue by jacobaustin123** [pytorch/pytorch#37738]
## Motivation

Currently, every call to `SummaryWriter.add_hparam` creates a new `SummaryWriter` which retains a separate record of the metrics associated with the given set of parameters. This makes it impossible to update the metric associated with a given set of hyperparameters. For example, I would like to do

```python
writer = SummaryWriter()
writer.add_hparams({"lr" : 0.001}, {"accuracy" : 0})
for epoch in range(epochs):
    ...
    writer.add_scalar("accuracy", get_accuracy(epoch), epoch)
```

and have the accuracy in the hparam console track the current "accuracy" scalar. Right now this fails because `add_hparams` creates a new directory with its own accuracy scalar that is always 0. This makes the "Show Metrics" button in Tensorboard useless, and it's also unexpected behavior because you would not expect a `SummaryWriter` to create a new `SummaryWriter` internally with a new directory for every parameter. If this behavior were changed, you could have results like [this](https://imgur.com/a/Z5u2Fer).

## Proposal

This is not trivial behavior to change. Tensorflow handles this by supporting

```python
  with tf.summary.create_file_writer(run_dir).as_default():
    hp.hparams(hparams)  # record the values used in this trial
    accuracy = train_test_model(hparams)
    tf.summary.scalar(METRIC_ACCURACY, accuracy, step=1)
```

but PyTorch's summary writer does not have a notion of `as_default()`, and it's not clear that `add_hparams` rightly belongs as a member function in `SummaryWriter`. I see three ways to support this.

1. Have `add_hparams` return the new `SummaryWriter` that it used to write the summary. This could be ignored if not needed but otherwise it could be used to update metrics during a training run. This has the downside of returning writers that need to be closed. A `return_writer=False` flag could be added to mitigate this.
2. Add a `use_current_writer=False` flag to `add_hparams` that would use the default file writer if `True` and track global metrics instead of creating a new `SummaryWriter`. This would be disabled by default, but would allow users to track global metrics if desired. 
3. Add a `replace_writer` flag to `add_hparams` that replaces the `file_writer` in the `SummaryWriter` with the new writer created by `add_hparams`. This would be False by default, but could be set to True if you want the metrics to track future calls to `add_scalar`.

I think the first is preferable, although it still pollutes the log directory with new files for every call to `add_hparams`. The third is a bit cleaner with respect to cleaning up file pointers, but it's more confusing from a user perspective.

## Code changes

First approach:

`torch/utils/tensorboard/writer.py:305`
```python
w_hp = SummaryWriter(log_dir=logdir)
w_hp.file_writer.add_summary(exp)
w_hp.file_writer.add_summary(ssi)
w_hp.file_writer.add_summary(sei)
for k, v in metric_dict.items():
    w_hp.add_scalar(k, v)

if return_writer:
    return w_hp
else:
    w_hp.close()
```

## Additional context

This is the kind of metric view you can access if this is supported.

<img width="500" alt="Screen Shot 2020-05-03 at 2 24 16 PM" src="https://user-images.githubusercontent.com/28993550/80922443-2f121800-8d4b-11ea-8a7c-1081f06b641c.png">

Currently, this is always a single value, i.e.

<img width="500" alt="Screen Shot 2020-05-03 at 2 38 04 PM" src="https://user-images.githubusercontent.com/28993550/80922521-b6f82200-8d4b-11ea-98ae-d607cdeafd8c.png">

**Comment by dreizehnutters** [comment]
I propose a trivial fix for this issue.

### Code to change
---
Idea: Add a `global_step` variable to the `add_scalar` call
`torch/utils/tensorboard/writer.py:266`
```
def add_hparams(
    self, hparam_dict, metric_dict, hparam_domain_discrete=None, run_name=None, global_step=None
):
```

`torch/utils/tensorboard/writer.py:308`
```
with SummaryWriter(log_dir=logdir) as w_hp:
    w_hp.file_writer.add_summary(exp)
    w_hp.file_writer.add_summary(ssi)
    w_hp.file_writer.add_summary(sei)
    for k, v in metric_dict.items():
        w_hp.add_scalar(k, v, global_step)
```

### Additional context
---
Using this fix allows to log metrics easily 

```
WRITER = SummaryWriter()
param_dict = {..}
exp_name = "..."
for epoch, data in enumerate(trainLoader):
    loss = model(data)
    WRITER.add_hparams(
          hparam_dict=param_dict,
          metric_dict={'_GLOBAL/loss':loss},
          run_name=exp_name,
          global_step=epoch)
```

This produces the following:
![image](https://i.imgur.com/kYD76jE.png)


What do you think @jacobaustin123

**Comment by dreizehnutters** [comment]
I submitted a pull request for this issue. PR #50653

**Comment by jleaves** [comment]
Any update on this?

**Comment by jbohnslav** [comment]
would really like this to be incorporated

**Comment by dreizehnutters** [comment]
The reviewer never commented on the PR.

**Comment by vwxyzjn** [comment]
Hey @dreizehnutters, I am not a pytorch reviewer, but I personally like `flax`'s tensorboard API a bit more (https://github.com/google/flax/blob/9901056b96fd629aec68422d5d70ceeb2e0195bb/flax/metrics/tensorboard.py#L200, https://github.com/google/flax/blob/9901056b96fd629aec68422d5d70ceeb2e0195bb/examples/ppo/ppo_lib.py#L295-L296)

Its usage looks like this:

```python
  summary_writer = tensorboard.SummaryWriter(model_dir)
  summary_writer.hparams(dict(config))
```

And it produces https://tensorboard.dev/experiment/pY7D2qYQQLO9ZT5lA9PFPA/#hparams

<img width="1642" alt="image" src="https://user-images.githubusercontent.com/5555347/167984876-11d1dd1d-bd95-4bd3-a222-3ce1db56c8d0.png">

**Comment by vwxyzjn** [comment]
Actually, my problem is solved based on https://github.com/lanpa/tensorboardX/issues/479#issuecomment-634438288.

I just needed to do the following


```
    writer = SummaryWriter(f"runs/{run_name}")
    from torch.utils.tensorboard.summary import hparams
    exp, ssi, sei = hparams(vars(args), metric_dict={"charts/episodic_return": 0})   
    writer.file_writer.add_summary(exp)                 
    writer.file_writer.add_summary(ssi)                 
    writer.file_writer.add_summary(sei)  
```

And it works!

<img width="1642" alt="image" src="https://user-images.githubusercontent.com/5555347/167987253-5813a4ed-4647-41da-93bc-5e09f2589833.png">

**Comment by ringohoffman** [comment]
@dreizehnutters @jbohnslav @jleaves @jacobaustin123 `global_step` was added to `add_hparams` via https://github.com/pytorch/pytorch/pull/109572!
